gtn/functions/compose.cpp

```diff
@@ -450,3 +451,665 @@ Graph compose(
 
 } // namespace detail
 } // namespace gtn
+
+namespace gtn {
+namespace detail {
+namespace dataparallel {
+// Change AOS to SOA
+struct GraphDataParallel {
+  std::vector<int> accept;
+  std::vector<int> start;
+
+  // Has one more element than the number of nodes
+  // One value per node - i-th value corresponds to i-th node
+  // Last element is the total number of arcs, so that
+  // each element and its neighbor forms a range 
+  std::vector<int> inArcOffset;
+  std::vector<int> outArcOffset;
+
+  // One value per arc
+  std::vector<int> inArcs;
+  std::vector<int> outArcs;
+
+  // One value per arc
+  // i-th value corresponds to i-th arc
+  std::vector<int> ilabels;
+  std::vector<int> olabels;
+  std::vector<int> srcNodes;
+  std::vector<int> dstNodes;
+  std::vector<float> weights;
+};
+
+namespace {
+// Exclusive/Inclusive prefix sum. The returned vector
```

Comment:
Exclusive sum is different from inclusive sum and
this function actually implements exclusive sum.

So the comment here `Exclusive/Inclusive` is misleading.